### ISSUE TITLE
you don't have to use coffeescript anymore

### ISSUE THREAD

**Issue by aeosynth** [ccd0/4chan-x#829]
it's 2016 and what is this

**Comment by ccd0** [comment]
i do want to migrate away from Coffeescript, but not enough that I'd rewrite everything just to do so. But I've been considering a gradual replacement.

**Comment by aeosynth** [comment]
in my new projects i use [babel](https://github.com/babel/babel) and make (instead of grunt)

**Comment by ccd0** [comment]
Speaking of grunt and make, needless recompilation is something I want to fix: #833

**Comment by aeosynth** [comment]
make automatically does that, looks at timestamps to figure out what needs compiling

**Comment by aeosynth** [comment]
makefile tutorial: https://gist.github.com/isaacs/62a2d1825d04437c6f08

**Comment by aeosynth** [comment]
there's a coffee -> es6 convervter: https://github.com/decaffeinate/decaffeinate. it fails easily, but it should help.

**Comment by r3c0d3x** [comment]
How should we go about converting it to ES6 without disrupting the current development? Maybe have it on a separate branch until it's completely converted?

**Comment by aeosynth** [comment]
i mean isn't there already a system where coffee and js get combined?

**Comment by ccd0** [comment]
@aeosynth
That should help if it's not too broken. I suspect some stuff will still have to be handled manually, but we could take it a few files at a time starting with modules we're changing and need to test anyway.

@r3c0d3x
I already have a few files in plain JS, and I can run the JS files through babel if needed. Although when browsers are ready I'd like to get to a point where there's no compilation step at all, just interpolating a few templates, linting, and concatenating.

It would also be worth seeing what @zixaphir and @vampiricwulf think of this since if Appchan X continues to be developed it would make merges and/or cherry-picking from/to 4chan X more difficult unless they also make the switch.

**Comment by aeosynth** [comment]
> I suspect some stuff will still have to be handled manually

more like, some stuff might not need too much manual handling

we're already at the point where native js is good enough. i'm using babel mainly for module + jsx syntax, neither of which are used in this project.

**Comment by ccd0** [comment]
As decaffeinate is right now, all uses of `a?()` and `a?.b` have to be converted manually (https://github.com/decaffeinate/decaffeinate/issues/176), as well as many instances of `for` loops used as expressions (https://github.com/decaffeinate/decaffeinate/issues/156). The lodash templates used throughout the codebase will have to be converted manually regardless of improvements in decaffeinate. And some of the code generated by decaffeinate appears to be broken. So this will definitely be a semi-manual process, but I think decaffeinate will expedite it.

**Comment by aeosynth** [comment]
you'll convert from your custom template syntax to template literals i assume?

also you could think about embracing the es6 module system, instead of making everything global. [rollup](https://github.com/rollup/rollup) is the bundler that i'm using, and it supports adding headers, etc.

**Comment by aeosynth** [comment]
babel's a generic compiler; you can use it just to compile templates

**Comment by ccd0** [comment]
> you'll convert from your custom template syntax to template literals i assume?

It's mostly just a matter of converting things like

    '<%= readBase64('ferongr.unreadDead.png') %>'

(https://github.com/ccd0/4chan-x/blob/1.11.35.2/src/Monitoring/Favicon.coffee#L15)
into

    `'<%= readBase64('ferongr.unreadDead.png') %>'`

at which point they should convert just fine. The ones like `<% if (type === 'crx') { %>` are more work, but not used as often.

> also you could think about embracing the es6 module system, instead of making everything global.

Probably going to do something like that, but I expect I'll first have to fix a bunch of circular dependencies.

**Comment by aeosynth** [comment]
you could look at the [inline environment variables transform](http://babeljs.io/docs/plugins/transform-inline-environment-variables/), or maybe make your own transform

modules should be fine with circular dependencies

i'm not familiar with the current state of the code, so my suggestions have to be vague

**Comment by ccd0** [comment]
That's good, although I want to get rid of the circular dependencies anyway since they make analyzing code changes more complicated.

A bigger concern I have about module bundlers is that if they take more than a second to run and/or have any chance of outputting broken code (and would therefore need to have their output linted), that breaks incremental builds. Rollup in particular looks like complexity I don't need, but a simpler bundler might be useful.

**Comment by aeosynth** [comment]
rollup is about the simplest module bundler out there; what complexity are you referring to? anything simpler would be the current method of simply concatenating files. any tool has a chance of containing bugs, but module bundlers are a common tool for production use.

**Comment by ccd0** [comment]
Looking at rollup some more, I definitely don't want to use it for 4chan X, because apparently the output would look something like this
http://bit.ly/1rjOwZD
with everything combined into one namespace, no indication of what module they came from, and `$1`, `$2`, etc. appended to names where there would be clashes. That's fine if you're only ever going to edit the original module files, but users will often want to make edits to the userscripts they've downloaded without bothering with the original source code. Unless there's some option that would change how the imported functions are named, this doesn't work for me.

**Comment by ccd0** [comment]
I suppose I could do this
http://bit.ly/1VTm9hG
but then I'm not sure what I'm gaining over just writing it that way in the first place and concatenating everything.

And even that starts to look ugly again if I start to define private functions:
http://bit.ly/1ULlKbL
whereas what I would do with concatenation is just stick `private_function` inside an IIFE where `Feature1` is defined.

**Comment by i-cant-git** [comment]
Thank you @ccd0 but you don't have to listen anybody when they have no contributions to the script.
I personally like CoffeeScript so much that I prefer it over TypeScript or ES6.

**Comment by ccd0** [comment]
Aeosynth started 4chan X, but if people currently contributing or interested in contributing prefer CoffeeScript, I could stick with that. But this isn't just something aeosynth wants, it's something I've wanted to do for a while.

I'm personally worried that the use of CoffeeScript is limiting potential contributors. I do see a lot of people saying that they would contribute if they knew CoffeeScript instead of just Javascript. I am a bit skeptical that anyone who couldn't pick up enough CoffeeScript to work on the project in a few hours would contribute very good code in any language, but even a small barrier can cause people to lose interest.

Another issue with CoffeeScript is that it adds complexity to the build process which can result in errors. These aren't necessarily bugs in CoffeeScript itself, but it's not uncommon for CoffeeScript to surprise people when converted to Javascript. Putting a for loop at the end of a function is an example of something that surprises most people. Arguably this is a matter of people being less proficient with CoffeeScript than with Javascript, but is anyone not more proficient with the latter? You have to know the latter to do the former.

Compilation also makes the build process slower. I'd prefer a build process that wasn't much more than concatenation. Switching to ES6 wouldn't allow us to drop compilation immediately if we want to keep supporting Pale Moon and other minor browsers, as we'd need to use babel until the [rebase](https://github.com/MoonchildProductions/Tycho) finishes, but it would definitely mean dropping the compilation step within about a year or so. If any forks don't support the ES6 we want to use by then, we should consider them dead and drop support for them.

**Comment by aeosynth** [comment]
any update on this? i was thinking converting files could be a neat warmup exercise.

**Comment by mahkoh** [comment]
>I'm personally worried that the use of CoffeeScript is limiting potential contributors.

If I may say so: One obstacle is the size of the repository. At over 400MB it takes more than one hour to do a full clone from github.

**Comment by ccd0** [comment]
@mahkoh
I think I could get it down to ~20 MB if I removed all the `builds/` stuff. Then that could go in a separate repository. But this involves rewriting history (although I would keep the old version in the other repository), so let's make this into a new issue for visibility's sake.

Alternately, if you don't need the full history, you can check out only recent history by adding `--depth 100` or even `--depth 1` to your `git clone` command line. This is mentioned in https://github.com/ccd0/4chan-x/blob/master/CONTRIBUTING.md.

**Comment by ccd0** [comment]
@mahkoh
New issue: #1107

**Comment by hails** [comment]
Hey, I'm considering to rewrite this using THE BEST OF THE JAVASCRIPT OF $CURRENT_YEAR
Jokes aside, I want to help to write everything using just plain ES6, maybe some webpack + babel (because you know, Javascript)...
Let's do it!

**Comment by aeosynth** [comment]
best js of current year is probably typescript

**Comment by ccd0** [comment]
I would be more interested in TypeScript if it could detect issues like this:
```
const dict1: { [x: string]: { [y: string]: string } } = {}
const remoteInput: string[] = ["__proto__", "polluted", "yes"]
dict1[remoteInput[0]][remoteInput[1]] = remoteInput[2]

const dict2: { [x: string]: boolean } = {}
const result: boolean = dict2.polluted
alert(result)
```

**Comment by saxamaphone69** [comment]
>issue opened 4 years ago

Damn, time does fly.

I assume the decaffeinate process has improved since then?
Realistically, what timeline/next steps/more needs to be done to move to away from coffeescript to ES6 (if that's still the [current](https://github.com/ccd0/4chan-x/issues/1473#issuecomment-518033276) consideration) or otherwise?

**Comment by ccd0** [comment]
It's just waiting on me right now. The switch and your pull request should be the next things to get looked at when I get around to making major changes again. I intend to update to Coffeescript 2 first which will switch the compiled output to ES6, then modules will have their source code changed into more natural-looking ES6 one by one, possibly with the help of decaffeinate, whenever they get major changes.

**Comment by saxamaphone69** [comment]
Happy to help and assist in any way, and as always, look forward to seeing 4chan X continue to develop/improve/add features and appreciate the effort you put in.

**Comment by aeosynth** [comment]
fyi es6 = es2015 and is now 5 years old. most of es2020 is supported by my browser (firefox), except for 2018 regex features

**Comment by aeosynth** [comment]
> Damn, time does fly.
everyone from the old crew is or is approaching boomerhood.

**Comment by saxamaphone69** [comment]
>boomerhood

My first userstyle for 4chan was 2009. 11 years later and I'm still doing the same shit. Boomer confirmed.

**Comment by aeosynth** [comment]
any updates? i see you're still using coffeescript 1.x

**Comment by TuxedoTako** [comment]
I started a fork to a js/ts migration. It's still work in progress, but I'm optimistic I can get it fully functional soon™. https://github.com/TuxedoTako/4chan-xt

**Comment by aeosynth** [comment]
it's 2026 and what is this
